spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/SparkBinPackPositionDeletesRewriter.java

```diff
@@ -125,10 +128,11 @@ private Dataset<Row> dataFiles(Types.StructType partitionType, StructLike partit
         IntStream.range(0, fields.size())
             .mapToObj(
                 i -> {
-                  Class<?> type = fields.get(i).type().typeId().javaClass();
-                  Object value = partition.get(i, type);
+                  Type type = fields.get(i).type();
+                  Object value = partition.get(i, type.typeId().javaClass());
+                  Object converted = SparkValueConverter.convertToSpark(type, value);
                   Column col = col("partition." + fields.get(i).name());
-                  return col.equalTo(value);
+                  return col.equalTo(lit(converted));
```

Comment:
Does this work correctly when the partition value is null? Do we have to use `eqNullSafe`?

Reply:
yep, done, and added test